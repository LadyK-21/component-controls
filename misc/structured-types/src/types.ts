import ts from 'typescript';

export interface JSDocTypeTag {
  type?: string;
  description?: string;
}

export interface JSDocExample {
  caption?: string;
  content?: string;
}

export interface JSDocPropTag {
  tag: string;
  content?: string;
}

export enum PropKind {
  Object = 0,
  String = 1,
  Number = 2,
  Boolean = 3,
  Union = 4,
  Enum = 5,
  Tuple = 6,
  Rest = 7,
  Undefined = 8,
  Unknown = 9,
  Null = 10,
  Function = 11,
  Void = 12,
  Class = 13,
  Interface = 14,
  Type = 15,
  Array = 16,
  Any = 17,
  Reference = 18,
  Predicate = 19,
  Index = 20,
  Constructor = 21,
  Getter = 22,
  Setter = 23,
  BigInt = 24,
  Component = 25,
}

export interface PropType {
  /**
   * generic properties
   */
  kind?: PropKind;
  displayName?: string;
  parent?: string;
  optional?: boolean;
  readonly?: boolean;
  abstract?: boolean;
  visibility?: 'private' | 'protected' | 'public';
  static?: boolean;
  type?: PropType | string;
  /**
   * framework name as set by plugins
   * ie 'react'...
   */
  framework?: string;
  /**
   * jsdoc comments parsing
   */
  description?: string;
  fires?: string[];
  see?: string[];
  examples?: JSDocExample[];
  tags?: JSDocPropTag[];
  summary?: string;
  deprecated?: string | true;
}

export interface ObjectProp extends PropType {
  kind: PropKind.Object;
  properties?: PropType[];
  value?: PropType[];
}

export const isObjectProp = (prop: PropType): prop is ObjectProp => {
  return prop.kind === PropKind.Object;
};

export interface StringProp extends PropType {
  kind: PropKind.String;
  value?: string;
}

export const isStringProp = (prop: PropType): prop is StringProp => {
  return prop.kind === PropKind.String;
};

export interface NumberProp extends PropType {
  kind: PropKind.Number;
  value?: number;
}

export const isNumberProp = (prop: PropType): prop is NumberProp => {
  return prop.kind === PropKind.Number || prop.kind === PropKind.BigInt;
};

export interface BooleanProp extends PropType {
  kind: PropKind.Boolean;
  value?: boolean;
}

export const isBooleanProp = (prop: PropType): prop is BooleanProp => {
  return prop.kind === PropKind.Boolean;
};

export interface UnionProp extends PropType {
  kind: PropKind.Union;
  properties?: PropType[];
  value?: any;
}

export const isUnionProp = (prop: PropType): prop is UnionProp => {
  return prop.kind === PropKind.Union;
};

export interface EnumProp extends PropType {
  properties?: PropType[];
}

export const isEnumProp = (prop: PropType): prop is EnumProp => {
  return prop.kind === PropKind.Enum;
};

export interface RestProp extends PropType {
  kind: PropKind.Rest;
}

export const isRestProp = (prop: PropType): prop is RestProp => {
  return prop.kind === PropKind.Rest;
};

export interface TupleProp extends PropType {
  kind: PropKind.Tuple;
  properties?: PropType[];
}

export const isTupleProp = (prop: PropType): prop is TupleProp => {
  return prop.kind === PropKind.Tuple;
};

export interface UndefinedProp extends PropType {
  kind: PropKind.Undefined;
  value?: undefined;
}

export const isUndefinedProp = (prop: PropType): prop is UndefinedProp => {
  return prop.kind === PropKind.Undefined;
};

export interface UnknownProp extends PropType {
  kind: PropKind.Unknown;
  value?: unknown;
}

export const isUnknownProp = (prop: PropType): prop is UnknownProp => {
  return prop.kind === PropKind.Unknown;
};

export interface NullProp extends PropType {
  kind: PropKind.Null;
  value?: null;
}

export const isNullProp = (prop: PropType): prop is NullProp => {
  return prop.kind === PropKind.Null;
};

export interface BaseFunctionProp extends PropType {
  parameters?: PropType[];
  returns?: PropType;
  types?: PropType[];
}

export interface FunctionProp extends BaseFunctionProp {
  kind: PropKind.Function;
}

export interface ConstructorProp extends BaseFunctionProp {
  kind: PropKind.Constructor;
}

export interface GetterProp extends BaseFunctionProp {
  kind: PropKind.Getter;
}
export interface SetterProp extends BaseFunctionProp {
  kind: PropKind.Getter;
}

export interface PredicateProp extends BaseFunctionProp {
  kind: PropKind.Predicate;
}

export const isFunctionProp = (prop: PropType): prop is FunctionProp => {
  return prop.kind === PropKind.Function;
};

export const isFunctionBaseType = (
  prop: PropType,
): prop is BaseFunctionProp => {
  return (
    prop.kind === PropKind.Function ||
    prop.kind === PropKind.Constructor ||
    prop.kind === PropKind.Getter ||
    prop.kind === PropKind.Setter ||
    prop.kind === PropKind.Predicate
  );
};

export interface VoidProp extends PropType {
  value?: void;
}

export const isVoidProp = (prop: PropType): prop is VoidProp => {
  return prop.kind === PropKind.Void;
};

export interface ClassProp extends PropType {
  implements?: InterfaceProp[];
  extends?: string[];
  generics?: PropType[];
  properties?: PropType[];
}
export interface ComponentProp extends ClassProp {
  kind: PropKind.Component;
}
export const isClassProp = (prop: PropType): prop is ClassProp => {
  return prop.kind === PropKind.Class;
};

export interface InterfaceProp extends PropType {
  extends?: string[];
  properties?: PropType[];
  generics?: PropType[];
}

export const isInterfaceProp = (prop: PropType): prop is InterfaceProp => {
  return prop.kind === PropKind.Interface;
};

export interface TypeProp extends PropType {
  extends?: string[];
  properties?: PropType[];
  generics?: PropType[];
}

export const isTypeProp = (prop: PropType): prop is TypeProp => {
  return prop.kind === PropKind.Type;
};

export type HasGenericsProp = TypeProp | InterfaceProp | ClassProp;

export const hasGenerics = (prop: PropType): prop is HasGenericsProp => {
  return (
    prop.kind === PropKind.Type ||
    prop.kind === PropKind.Class ||
    prop.kind === PropKind.Interface ||
    prop.kind === PropKind.Array
  );
};

export interface ArrayProp extends PropType {
  properties?: PropType[];
  value?: PropType[];
}

export const isArrayProp = (prop: PropType): prop is ArrayProp => {
  return prop.kind === PropKind.Array;
};

export interface AnyProp extends PropType {
  value?: any;
}

export const isAnyProp = (prop: PropType): prop is AnyProp => {
  return prop.kind === PropKind.Any;
};

export interface ReferenceProp extends PropType {
  types?: PropType[];
}

export const isReferenceProp = (prop: PropType): prop is ReferenceProp => {
  return prop.kind === PropKind.Reference;
};

export const isPredicateProp = (prop: PropType): prop is PredicateProp => {
  return prop.kind === PropKind.Predicate;
};

export interface IndexProp extends PropType {
  index: PropType;
  properties: PropType[];
}

export const isIndexProp = (prop: PropType): prop is IndexProp => {
  return prop.kind === PropKind.Index;
};

export type ValueProp =
  | AnyProp
  | ArrayProp
  | VoidProp
  | UnionProp
  | NullProp
  | UnknownProp
  | UndefinedProp
  | BooleanProp
  | NumberProp
  | StringProp
  | ObjectProp;

export const isValueProp = (prop: PropType): prop is ValueProp => {
  return (
    prop.kind === PropKind.Any ||
    prop.kind === PropKind.Array ||
    prop.kind === PropKind.Void ||
    prop.kind === PropKind.Union ||
    prop.kind === PropKind.Null ||
    prop.kind === PropKind.Unknown ||
    prop.kind === PropKind.Undefined ||
    prop.kind === PropKind.Boolean ||
    prop.kind === PropKind.Number ||
    prop.kind === PropKind.String ||
    prop.kind === PropKind.Object
  );
};

export type ClassLikeProp = ClassProp | InterfaceProp | TypeProp;

export const isClassLikeProp = (prop: PropType): prop is ClassLikeProp => {
  return (
    prop.kind === PropKind.Class ||
    prop.kind === PropKind.Interface ||
    prop.kind === PropKind.Type
  );
};
export type ObjectLikeProp = ClassLikeProp | EnumProp | ObjectProp | IndexProp;

export const isObjectLikeProp = (prop: PropType): prop is ObjectLikeProp => {
  return (
    isClassLikeProp(prop) ||
    prop.kind === PropKind.Index ||
    prop.kind === PropKind.Enum ||
    prop.kind === PropKind.Object
  );
};

export type PropDiagnostic = {
  category: ts.DiagnosticCategory;
  message: string;
  row?: number;
  column?: number;
  fileName?: string;
};
export type PropTypes = {
  [propName: string]: PropType;
} & { __parents?: Record<string, PropType>; __diagnostics?: PropDiagnostic[] };

export const typeNameToPropKind = (type: string): PropKind | undefined => {
  const loopup: Record<string, PropKind> = {
    object: PropKind.Object,
    string: PropKind.String,
    number: PropKind.Number,
    boolean: PropKind.Boolean,
    union: PropKind.Union,
    Enum: PropKind.Enum,
    Tuple: PropKind.Tuple,
    function: PropKind.Function,
    class: PropKind.Function,
    type: PropKind.Function,
    array: PropKind.Array,
  };
  return loopup[type];
};

export type JSDocInfoType = {
  comment?: ts.JSDocTag['comment'];
};

export const propValue = (prop: PropType, value?: string): PropType => {
  if (value) {
    if (isNumberProp(prop)) {
      prop.value = Number(value);
    } else if (isBooleanProp(prop)) {
      prop.value = Boolean(value);
    } else {
      (prop as StringProp).value = value;
    }
  }
  return prop;
};
