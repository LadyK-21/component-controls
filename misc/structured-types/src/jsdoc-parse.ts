import { Node } from '@babel/types';
import deepmerge from 'deepmerge';
import { PropType } from './types';
import { jsdocCommentToMember } from './jsdoc/jsdocCommentToMember';

export const getNodeComments = (node: Node): string | undefined =>
  node.leadingComments
    ? node.leadingComments
        .filter(comment => comment.type === 'CommentBlock')
        .map(comment => comment.value)
        .join('/n')
    : undefined;
export const mergeJSDocComments = (
  prop: PropType = {},
  comments?: string,
): PropType | null => {
  if (comments) {
    const parsed = jsdocCommentToMember(comments);
    if (parsed === null) {
      return null;
    }
    if (parsed) {
      const merged = deepmerge<PropType>(prop, parsed, {
        clone: false,
        arrayMerge: (dest: any[], src: any[]) => {
          const result =
            dest.length > 0
              ? dest.map((s, idx) => {
                  const existingIdx =
                    src.length === dest.length
                      ? idx
                      : src.findIndex(
                          p => p.displayName === (s as PropType).displayName,
                        );
                  if (existingIdx >= 0) {
                    Object.assign(s, {
                      ...src[existingIdx],
                      ...s,
                    });
                  }
                  return s;
                })
              : dest;
          return result;
        },
      });
      return merged;
    }
  }
  return prop;
};
