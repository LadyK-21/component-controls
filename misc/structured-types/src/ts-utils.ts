import * as ts from 'typescript';
import { PropKind, PropType } from './types';

type VariableDeclaration =
  | ts.VariableDeclaration
  | ts.ParameterDeclaration
  | ts.BindingElement
  | ts.PropertyDeclaration
  | ts.PropertySignature
  | ts.JsxAttribute
  | ts.EnumMember;

export const isVariableLikeDeclaration = (
  node: ts.Node,
): node is VariableDeclaration => {
  return (
    node.kind === ts.SyntaxKind.VariableDeclaration ||
    node.kind === ts.SyntaxKind.Parameter ||
    node.kind === ts.SyntaxKind.BindingElement ||
    node.kind === ts.SyntaxKind.PropertyDeclaration ||
    node.kind === ts.SyntaxKind.PropertyAssignment ||
    node.kind === ts.SyntaxKind.PropertySignature ||
    node.kind === ts.SyntaxKind.JsxAttribute ||
    node.kind === ts.SyntaxKind.EnumMember
  );
};

export const isHasType = (node: ts.Node): node is ts.HasType => {
  return (
    node.kind === ts.SyntaxKind.CallSignature ||
    node.kind === ts.SyntaxKind.ConstructSignature ||
    node.kind === ts.SyntaxKind.MethodSignature ||
    node.kind === ts.SyntaxKind.IndexSignature ||
    node.kind === ts.SyntaxKind.FunctionType ||
    node.kind === ts.SyntaxKind.MethodDeclaration ||
    node.kind === ts.SyntaxKind.Constructor ||
    node.kind === ts.SyntaxKind.GetAccessor ||
    node.kind === ts.SyntaxKind.SetAccessor ||
    node.kind === ts.SyntaxKind.FunctionExpression ||
    node.kind === ts.SyntaxKind.ArrowFunction ||
    node.kind === ts.SyntaxKind.VariableDeclaration ||
    node.kind === ts.SyntaxKind.Parameter ||
    node.kind === ts.SyntaxKind.PropertySignature ||
    node.kind === ts.SyntaxKind.PropertyDeclaration ||
    node.kind === ts.SyntaxKind.TypePredicate ||
    node.kind === ts.SyntaxKind.ParenthesizedType ||
    node.kind === ts.SyntaxKind.TypeOperator ||
    node.kind === ts.SyntaxKind.MappedType ||
    node.kind === ts.SyntaxKind.TypeAssertionExpression ||
    node.kind === ts.SyntaxKind.AsExpression ||
    node.kind === ts.SyntaxKind.TypeAliasDeclaration
  );
};

export const tsKindToPropKind: { [key in ts.SyntaxKind]?: PropKind } = {
  [ts.SyntaxKind.StringKeyword]: PropKind.String,
  [ts.SyntaxKind.StringLiteral]: PropKind.String,
  [ts.SyntaxKind.NumberKeyword]: PropKind.Number,
  [ts.SyntaxKind.NumericLiteral]: PropKind.Number,
  [ts.SyntaxKind.BooleanKeyword]: PropKind.Boolean,
  [ts.SyntaxKind.EnumDeclaration]: PropKind.Enum,
  [ts.SyntaxKind.UnionType]: PropKind.Union,
  [ts.SyntaxKind.ClassDeclaration]: PropKind.Class,
  [ts.SyntaxKind.ClassExpression]: PropKind.Class,
  [ts.SyntaxKind.ClassDeclaration]: PropKind.Class,
  [ts.SyntaxKind.InterfaceDeclaration]: PropKind.Interface,
  [ts.SyntaxKind.TypeLiteral]: PropKind.Type,
  [ts.SyntaxKind.TypeReference]: PropKind.Type,
  [ts.SyntaxKind.CallSignature]: PropKind.Function,
  [ts.SyntaxKind.ConstructSignature]: PropKind.Constructor,
  [ts.SyntaxKind.MethodSignature]: PropKind.Function,
  [ts.SyntaxKind.FunctionDeclaration]: PropKind.Function,
  [ts.SyntaxKind.FunctionType]: PropKind.Function,
  [ts.SyntaxKind.MethodDeclaration]: PropKind.Function,
  [ts.SyntaxKind.Constructor]: PropKind.Constructor,
  [ts.SyntaxKind.GetAccessor]: PropKind.Getter,
  [ts.SyntaxKind.SetAccessor]: PropKind.Setter,
  [ts.SyntaxKind.FunctionExpression]: PropKind.Function,
  [ts.SyntaxKind.ArrowFunction]: PropKind.Function,
  [ts.SyntaxKind.TypePredicate]: PropKind.Type,
  [ts.SyntaxKind.TypeOperator]: PropKind.Type,
  [ts.SyntaxKind.MappedType]: PropKind.Type,
  [ts.SyntaxKind.TypeAssertionExpression]: PropKind.Type,
  [ts.SyntaxKind.TypeAliasDeclaration]: PropKind.Type,
  [ts.SyntaxKind.IndexSignature]: PropKind.Index,
  [ts.SyntaxKind.ArrayType]: PropKind.Array,
};
export type ObjectTypeDeclaration =
  | ts.ClassDeclaration
  | ts.ClassExpression
  | ts.InterfaceDeclaration;
export const isObjectTypeDeclaration = (
  node: ts.Node,
): node is ObjectTypeDeclaration => {
  return (
    node.kind === ts.SyntaxKind.ClassDeclaration ||
    node.kind === ts.SyntaxKind.ClassExpression ||
    node.kind === ts.SyntaxKind.InterfaceDeclaration
  );
};

export type TypeParameterType = ObjectTypeDeclaration | ts.TypeAliasDeclaration;
export const isTypeParameterType = (node: ts.Node): node is TypeParameterType =>
  isObjectTypeDeclaration(node) ||
  node.kind === ts.SyntaxKind.TypeAliasDeclaration;

export type GenericsType = TypeParameterType | ts.TypeLiteralNode;

export const isGenericsType = (node: ts.Node): node is GenericsType =>
  isTypeParameterType(node) || node.kind === ts.SyntaxKind.TypeLiteral;

export type FunctionLike =
  | ts.FunctionDeclaration
  | ts.ConstructorDeclaration
  | ts.GetAccessorDeclaration
  | ts.SetAccessorDeclaration
  | ts.ArrowFunction
  | ts.FunctionTypeNode
  | ts.MethodSignature
  | ts.MethodDeclaration;

export const isFunctionLike = (node: ts.Node): node is FunctionLike => {
  return (
    node.kind === ts.SyntaxKind.FunctionDeclaration ||
    node.kind === ts.SyntaxKind.ArrowFunction ||
    node.kind === ts.SyntaxKind.Constructor ||
    node.kind === ts.SyntaxKind.GetAccessor ||
    node.kind === ts.SyntaxKind.SetAccessor ||
    node.kind === ts.SyntaxKind.FunctionType ||
    node.kind === ts.SyntaxKind.MethodSignature ||
    node.kind === ts.SyntaxKind.MethodDeclaration
  );
};

export type ArrayLike =
  | ts.ArrayTypeNode
  | ts.ArrayLiteralExpression
  | ts.TypeReferenceNode;

export const isArrayLike = (node: ts.Node): node is ArrayLike => {
  return (
    node.kind === ts.SyntaxKind.ArrayType ||
    node.kind === ts.SyntaxKind.ArrayLiteralExpression ||
    (ts.isTypeReferenceNode(node) &&
      ['ArrayConstructor', 'Array'].includes(node.typeName.getText()))
  );
};

export const tsDefaults: ts.CompilerOptions = {
  jsx: ts.JsxEmit.ReactJSX,
  module: ts.ModuleKind.CommonJS,
  target: ts.ScriptTarget.ES2017,
  noImplicitAny: true,
  noImplicitReturns: true,
  strictNullChecks: true,
  strictFunctionTypes: true,
  strictBindCallApply: true,
  strictPropertyInitialization: true,
  noImplicitThis: true,
  alwaysStrict: true,
  allowJs: true,
  checkJs: true,
};

export type CompileOptions = {
  tsOptions?: ts.CompilerOptions & { lang?: 'typescript' | 'javascript' };
};

/**
 * parsing options
 */
export interface ParseOptions {
  /**
   * type resolving custom function
   * ie from a react component will return the props type
   */
  resolvers?: TypeResolver[];
  /**
   * internal types - libs
   * by default includes classes such as `String`, `Function`...
   */
  internalTypes?: string[];
  /**
   * list of export names to be extracted.
   * by default all exports are extracted
   */
  extractNames?: string[];
  /**
   * whether to save the "parent" props
   * if set to false will result in a smaller result
   */
  saveParentProps?: boolean;
  /**
   * whether to collect errors/diagnostics
   */
  collectDiagnostics?: boolean;
}

export const defaultParseOptions: ParseOptions = {
  saveParentProps: true,
  internalTypes: [
    'Function',
    'CallableFunction',
    'NewableFunction',
    'String',
    'Boolean',
    'Booleanish',
    'Number',
    'Array',
    'Promise',
    'ConcatArray',
    'ReadonlyArray',
    'TemplateStringsArray',
  ],
};

export type DocsOptions = CompileOptions & ParseOptions;

export type TypeResolver = (props: {
  symbolType: ts.Type;
  declaration: ts.Declaration;
  checker: ts.TypeChecker;
}) => { type: ts.Type | undefined; intializer?: ts.Node; name?: string };

export const resolveType: (
  props: Parameters<TypeResolver>[0],
  resolvers?: TypeResolver[],
) => ReturnType<TypeResolver> = (props, resolvers) => {
  if (resolvers) {
    return resolvers.reduce(
      (acc: ReturnType<TypeResolver>, resolver) => {
        return (
          resolver({ ...props, symbolType: acc.type || props.symbolType }) ||
          acc
        );
      },
      { type: props.symbolType },
    );
  }
  return { type: props.symbolType };
};

export const getSymbolType = (
  checker: ts.TypeChecker,
  symbol: ts.Symbol,
): ts.Type | undefined => {
  const declaration = symbol.valueDeclaration || symbol.declarations?.[0];
  if (declaration) {
    const type = checker.getTypeOfSymbolAtLocation(symbol, declaration);
    if (
      !('intrinsicName' in type) ||
      ((type as unknown) as { intrinsicName: string }).intrinsicName !== 'error'
    ) {
      return type;
    }
  }
  const symbolType = checker.getDeclaredTypeOfSymbol(symbol);
  if (
    !('intrinsicName' in symbolType) ||
    ((symbolType as unknown) as { intrinsicName: string }).intrinsicName !==
      'error'
  ) {
    return symbolType;
  }
  return undefined;
};

export interface ISymbolParser {
  parseType(prop: PropType, node?: ts.Node): PropType;
  parseSymbol(symbol: ts.Symbol): PropType | null;
}

export const getInitializer = (
  declaration: ts.Node,
): ts.Expression | undefined =>
  isVariableLikeDeclaration(declaration)
    ? declaration?.initializer
    : ts.isBinaryExpression(declaration.parent)
    ? declaration.parent.right
    : undefined;
