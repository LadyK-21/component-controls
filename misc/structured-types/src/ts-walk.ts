import * as ts from 'typescript';
import { PropTypes, PropDiagnostic } from './types';
import { tsDefaults, DocsOptions } from './ts-utils';
import { SymbolParser } from './SymbolParser';

export const anaylizeFiles = (
  fileNames: string[],
  options: DocsOptions = {},
  host?: ts.CompilerHost,
): PropTypes => {
  const { tsOptions = tsDefaults, ...parseOptions } = options;
  const { extractNames, collectDiagnostics } = parseOptions || {};
  const program = ts.createProgram(fileNames, tsOptions, host);
  // Get the checker, we will use it to find more about classes
  const checker = program.getTypeChecker();

  const parser = new SymbolParser(checker, parseOptions);
  const parsed: PropTypes = {};
  for (const fileName of fileNames) {
    const sourceFile = program.getSourceFile(fileName);
    if (sourceFile) {
      const module = checker.getSymbolAtLocation(sourceFile);
      if (module) {
        const exports = checker.getExportsOfModule(module);
        exports.forEach(e => {
          const symbolName = e.getName();
          if (!extractNames || extractNames.includes(symbolName)) {
            const prop = parser.parseSymbol(e);
            if (prop) {
              parsed[symbolName] = prop;
            }
          }
        });
      }
    }
  }
  if (Object.keys(parser.parents).length) {
    parsed.__parents = parser.parents;
  }
  if (collectDiagnostics) {
    const allDiagnostics = ts
      .getPreEmitDiagnostics(program)
      .filter(({ file }) => file && fileNames.includes(file.fileName));
    if (allDiagnostics.length) {
      parsed.__diagnostics = allDiagnostics.map(
        ({ category, messageText, file, start }) => {
          const message =
            typeof messageText === 'string'
              ? messageText
              : messageText.messageText;

          const result: PropDiagnostic = {
            category,
            message,
          };
          if (file) {
            result.fileName = file.fileName
              .split('\\')
              .pop()
              ?.split('/')
              .pop();
            if (typeof start !== 'undefined') {
              const location = file.getLineAndCharacterOfPosition(start);
              result.row = location.line + 1;
              result.column = location.character + 1;
            }
          }
          return result;
        },
      );
    }
  }

  return parsed;
};
