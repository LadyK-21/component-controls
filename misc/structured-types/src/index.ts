import { getTypescriptConfig } from '@component-controls/typescript-config';
import { anaylizeFiles } from './ts-walk';
import { PropTypes } from './types';
import { tsDefaults, DocsOptions } from './ts-utils';
export * from './types';
export * from './ts-utils';

export const parseFiles = (
  filePaths: string[],
  options: DocsOptions = {},
): PropTypes => {
  if (!filePaths.length) {
    throw new Error('You need to supply at least one file');
  }
  options.tsOptions = {
    ...tsDefaults,
    ...getTypescriptConfig(filePaths[0], options.tsOptions),
  };
  const results = anaylizeFiles(filePaths, options);
  return results;
};
