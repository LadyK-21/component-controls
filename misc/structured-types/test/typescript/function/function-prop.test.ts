import path from 'path';
import { parseFiles } from '../../../src/index';

describe('function', () => {
  it('react fc', () => {
    const results = parseFiles([path.resolve(__dirname, 'react-fc.ts')]);
    expect(results).toMatchSnapshot();
  });
  it('function properties', () => {
    const results = parseFiles([path.resolve(__dirname, 'function-props.ts')]);

    expect(results).toEqual({
      fn: {
        displayName: 'fn',
        kind: 11,
        properties: [
          {
            kind: 1,
            description: 'custom property for a function',
            displayName: 'customProp',
            value: 'my custom prop',
          },
        ],
        returns: {
          kind: 12,
        },
      },
    });
  });

  it('object parameter', () => {
    const results = parseFiles([
      path.resolve(__dirname, 'object-parameter.ts'),
    ]);

    expect(results).toEqual({
      printCoord: {
        parameters: [
          {
            displayName: 'pt',
            description: 'object parameter',
            kind: 15,
            properties: [
              {
                description: 'x coordinate',
                kind: 2,
                displayName: 'x',
              },
              {
                description: 'optional y coordinate',
                kind: 2,
                optional: true,
                displayName: 'y',
              },
            ],
          },
        ],
        description: 'print coordinates',
        displayName: 'printCoord',
        kind: 11,
        returns: {
          kind: 12,
        },
      },
    });
  });
  it('generic function parameter', () => {
    const results = parseFiles([
      path.resolve(__dirname, 'generic-parameter.ts'),
    ]);
    expect(results).toEqual({
      genericFunction: {
        displayName: 'genericFunction',
        kind: 11,
        parameters: [
          {
            displayName: 'box',
            kind: 14,
            properties: [
              {
                parent: 'GenericInterface',
                displayName: 'm',
                kind: 15,
                type: 'T',
              },
            ],
            generics: [
              {
                displayName: 'T',
              },
            ],
          },
          {
            displayName: 'newContents',
            kind: 15,
            type: 'Type',
          },
        ],
        returns: {
          kind: 14,
          properties: [
            {
              displayName: 'm',
              kind: 15,
              type: 'T',
            },
          ],
          generics: [
            {
              displayName: 'T',
            },
          ],
          displayName: 'GenericInterface',
        },
        types: [
          {
            displayName: 'Type',
          },
        ],
      },
      __parents: {
        GenericInterface: {
          displayName: 'GenericInterface',
          kind: 14,
          properties: [
            {
              displayName: 'm',
              kind: 15,
              type: 'T',
            },
          ],
          generics: [
            {
              displayName: 'T',
            },
          ],
        },
      },
    });
  });

  it('extended parameter', () => {
    const results = parseFiles([
      path.resolve(__dirname, 'extends-parameter.ts'),
    ]);
    expect(results).toEqual({
      paintHomeyBear: {
        displayName: 'paintHomeyBear',
        kind: 11,
        parameters: [
          {
            kind: 15,
            properties: [
              {
                parent: 'T',
                kind: 1,
                displayName: 'm',
                description: 'base type member property',
              },
              {
                kind: 3,
                displayName: 'honey',
                description: 'own member',
              },
            ],
            description: 'extended type',
            displayName: 'ExtendT',
          },
        ],
        returns: {
          kind: 14,
          properties: [
            {
              kind: 3,
              displayName: 'honey',
              description: 'boolean type member',
            },
            {
              parent: 'Internal',
              kind: 1,
              displayName: 'm',
              description: 'string type member',
            },
          ],
          description: 'interface extending another one',
          displayName: 'Bear',
        },
        description: 'exported function',
      },
      __parents: {
        T: {
          displayName: 'T',
          kind: 15,
          properties: [
            {
              kind: 1,
              displayName: 'm',
              description: 'base type member property',
            },
          ],
          description: 'base type',
        },
        Internal: {
          displayName: 'Internal',
          kind: 14,
          properties: [
            {
              kind: 1,
              displayName: 'm',
              description: 'string type member',
            },
          ],
          description: 'internal interface with one member',
        },
      },
    });
  });

  it('tuple parameters', () => {
    const results = parseFiles([path.resolve(__dirname, 'tuple-parameter.ts')]);
    expect(results).toEqual({
      distanceFromOrigin: {
        displayName: 'distanceFromOrigin',
        kind: 11,
        parameters: [
          {
            kind: 6,
            properties: [
              {
                kind: 2,
              },
              {
                kind: 2,
              },
            ],
          },
        ],
        returns: {
          kind: 2,
        },
      },
    });
  });
  it('spread tuple parameters', () => {
    const results = parseFiles([
      path.resolve(__dirname, 'spread-tuple-parameter.ts'),
    ]);
    expect(results).toEqual({
      spreadTupleFunction: {
        displayName: 'spreadTupleFunction',
        kind: 11,
        parameters: [
          {
            displayName: 'args',
            kind: 6,
            properties: [
              {
                kind: 1,
              },
              {
                kind: 2,
              },
              {
                kind: 7,
                type: {
                  kind: 16,
                  properties: [
                    {
                      kind: 3,
                    },
                  ],
                },
              },
            ],
          },
        ],
        returns: {
          kind: 12,
        },
      },
    });
  });

  it('union parameter', () => {
    const results = parseFiles([path.resolve(__dirname, 'union-parameter.ts')]);
    expect(results).toEqual({
      printId: {
        displayName: 'printId',
        kind: 11,
        parameters: [
          {
            displayName: 'id',
            kind: 4,
            properties: [
              {
                kind: 2,
              },
              {
                kind: 1,
              },
            ],
          },
        ],
        returns: {
          kind: 12,
        },
      },
    });
  });

  it('jsdoc parameter', () => {
    const results = parseFiles([path.resolve(__dirname, 'jsdoc-parameter.ts')]);
    expect(results).toEqual({
      greet: {
        parameters: [
          {
            displayName: 'name',
            description: 'string type parameters',
            kind: 1,
          },
        ],
        description: 'greeting function',
        displayName: 'greet',
        kind: 11,
      },
    });
  });
  it('arrow fn jsdoc', () => {
    const results = parseFiles([path.resolve(__dirname, 'arrow-function.ts')]);
    expect(results).toEqual({
      arrowGreet: {
        description: 'arrow greeting function',
        displayName: 'arrowGreet',
        kind: 11,
        parameters: [
          {
            description: 'name parameter inline description',
            kind: 1,
            displayName: 'name',
          },
        ],
        returns: {
          kind: 12,
        },
      },
    });
  });
});
