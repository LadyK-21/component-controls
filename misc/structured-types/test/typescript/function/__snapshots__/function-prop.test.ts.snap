// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`function react fc 1`] = `
Object {
  "TypedInitializedFunction": Object {
    "displayName": "TypedInitializedFunction",
    "generics": Array [
      Object {
        "displayName": "P",
      },
    ],
    "kind": 11,
    "parameters": Array [
      Object {
        "displayName": "props",
        "generics": Array [
          Object {
            "displayName": "P",
          },
        ],
        "kind": 15,
        "properties": Array [
          Object {
            "displayName": "m",
            "kind": 1,
            "optional": true,
            "parent": "Props",
          },
          Object {
            "displayName": "children",
            "kind": 15,
            "optional": true,
            "parent": "PropsWithChildren",
            "type": "ReactNode",
          },
        ],
      },
    ],
    "properties": Array [
      Object {
        "displayName": "propTypes",
        "generics": Array [
          Object {
            "displayName": "P",
            "kind": 15,
          },
        ],
        "kind": 15,
        "optional": true,
        "parent": "FunctionComponent",
        "type": "WeakValidationMap<P>",
      },
      Object {
        "displayName": "contextTypes",
        "generics": Array [
          Object {
            "kind": 17,
          },
        ],
        "kind": 15,
        "optional": true,
        "parent": "FunctionComponent",
        "type": "ValidationMap<any>",
      },
      Object {
        "displayName": "defaultProps",
        "generics": Array [
          Object {
            "displayName": "P",
            "kind": 15,
          },
        ],
        "kind": 15,
        "optional": true,
        "parent": "FunctionComponent",
        "type": "Partial<P>",
      },
      Object {
        "displayName": "displayName",
        "kind": 1,
        "optional": true,
        "parent": "FunctionComponent",
      },
    ],
  },
  "__parents": Object {
    "FunctionComponent": Object {
      "displayName": "FunctionComponent",
      "generics": Array [
        Object {
          "displayName": "P",
        },
      ],
      "kind": 14,
      "properties": Array [
        Object {
          "displayName": "propTypes",
          "generics": Array [
            Object {
              "displayName": "P",
              "kind": 15,
            },
          ],
          "kind": 15,
          "optional": true,
          "type": "WeakValidationMap<P>",
        },
        Object {
          "displayName": "contextTypes",
          "generics": Array [
            Object {
              "kind": 17,
            },
          ],
          "kind": 15,
          "optional": true,
          "type": "ValidationMap<any>",
        },
        Object {
          "displayName": "defaultProps",
          "generics": Array [
            Object {
              "displayName": "P",
              "kind": 15,
            },
          ],
          "kind": 15,
          "optional": true,
          "type": "Partial<P>",
        },
        Object {
          "displayName": "displayName",
          "kind": 1,
          "optional": true,
        },
      ],
    },
    "Props": Object {
      "displayName": "Props",
      "kind": 14,
      "properties": Array [
        Object {
          "displayName": "m",
          "kind": 1,
          "optional": true,
        },
      ],
    },
    "PropsWithChildren": Object {
      "displayName": "PropsWithChildren",
      "generics": Array [
        Object {
          "displayName": "P",
        },
      ],
      "kind": 15,
      "properties": Array [
        Object {
          "displayName": "children",
          "kind": 15,
          "optional": true,
          "type": "ReactNode",
        },
      ],
    },
  },
}
`;
